proposals/init.md

```diff
@@ -0,0 +1,490 @@
+};
+s.LastName = "Parsons"; // Error: LastName is not settable
+```
+
+The rules around when `init` accessors are settable extend across type
+hierarchies. If the member is accessible and the object is known to be in the
+construction phase then the member is settable. That specifically allows for
+the following:
+
+```cs
+class Base
+{
+    public bool Value { get; init; }
+}
+
+class Derived : Base
+{
+    Derived()
+    {
+        // Not allowed with get only properties but allowed with init
+        Value = true;
+    }
+}
+
+class Consumption
+{
+    void Example()
+    {
+        var d = new Derived() { Value = true; };
+    }
+}
+
+```
+
+At the point a `init` accessor is invoked the instance is known to be 
+in the open construction phase. Hence an `init` accessor is allowed to take 
+the following actions:
+1. Call other `init` accessors available through `this`
+1. Assign `readonly` fields declared on the same type
+
+```cs
+class Complex
+{
+    readonly int Field1;
+    int Field2;
+    int Prop1 { get; init ; }
+    int Prop2
+    {
+        get => 42;
+        set
+        {
+            Field1 = 13; // okay
+            Field2 = 13; // okay
+            Prop1 = 13; // okay
+        }
+    }
+}
+```
+
+The ability to assign fields from an `init` accessor is limited to fields
+declared on the same type as the accessor. It cannot be used to assign fields
+in a base type. This rule ensures that type authors remain in control over the
+mutability behavior of their type. Developers who do not wish to utilize 
+`init` cannot be impacted from other types choosing to do so:
+
+```cs
+class Base
+{
+    internal readonly int Field;
+    internal int Property
+    {
+        get => Field;
+        init => Field = value; // Okay
+    }
+}
+
+class Derived : Base
+{
+    internal readonly int DerivedField;
+    internal int DerivedProperty
+    {
+        get => DerivedField;
+        init
+        {
+            DerivedField = 42; // Okay
+            Field = 13; // Error Field is readonly
+        }
+    }
+
+    public Derived()
+    {
+        Property = 42; // Okay 
+        Field = 13; // Error Field is readonly
+    }
+}
+```
+
+When `init` is used in a virtual property then all the overrides must also
+be marked as `init`. Likewise it is not possible to override a simple 
+`set` with `init`.
+
+An `interface` declaration can also particpate in `init` style initalization 
+via the following pattern:
+
+```cs
+interface IPerson
+{
+    string Name { get; init; }
+}
+
+class Init
+{
+    void M<T>() where T : IPerson, new()
+    {
+        var local = new T()
+        {
+            Name = "Jared"
+        };
+        local.Name = "Jraed"; // Error
+    }
+}
+```
+
+Restrictions of this feature:
+- The `init` accessor can only be used instance properties
+- A property cannot contain both an `init` and `set` accessor
+- All overrides of a property must have `init` if the base had `init`. This rule
+also applies to interface implementation.
```

Comment:
Interface implementations today allow you to only specify `get`, but add a `set` in the implementation. I think it's work allowing the same thing for specifying `init` in the interface but allowing `set` in the implementation.

Reply:
I can definitely get the motivation for allowing a class to implement a `get` only property with `get; init;`. That follows from `set` and seems pretty reasonable. I'm less certain about allowing a `get; init;` property be implemented with `get; set;`. It would require us to generate an explicit implementation of the interface under the hood (modreq would break the signature match). That makes me hesitate. I think I'd need to see some user feedback to want to take that on.